MarlinFirmware/Marlin
# Improve Delta probing / calibration

### Description

Fix a variety of issues with Delta probing:

1. Build error if no probe defined (Issue #15824)
2. ABL Bilinear G29 ignores probe offsets configured with M851
3. G33 ignores probe offsets configured with M851
4. G33 uses the reachable probe radius with nozzle-relative probes. This can cause damaging crashes even when the probe offset is configured in the header file.

(The following change was already implemented in 3f4636e916f0c2635f83f457fa662b90fc3a8809)
~~Also a tangentially related issue for some non-delta printers, with the same root-cause as #2. I did not explore this further or physically test the change, but the change seemed clearly needed due to its similarity to delta issues I encountered.~~

~~5. (Not delta) M851 changes ignored in MESH_(MIN|MAX)_(X|Y) for some configurations~~

### Benefits

- Allows dynamic configuration of probe offsets for delta, as for other motion systems.
- Prevents crashing if ignoring offsets would have probed off of bed
- G33 calibration is performed in an even circle around the bed, rather than skewed to one side due to the ignore probe offset.

### Behavior Change (Possible risk, seeking feedback)

One significant change of this is that probing during G33 is now probe-relative instead of nozzle relative. This appears to work fine on my printer, but I do not have the experience with the G33 calibration mechanism to know whether this will cause inefficient corrections. The probe points are now evenly distributed around the center of the bed for any probe offset, but the relationship of the arms to towers during probing has changed from previous behavior. Perhaps someone like @AnHardt knows more about this and can voice an opinion?

### Testing Methodology

I validated this on an AnyCubic Kossel Linear Plus. This uses a magnetically mounted fixed probe, which is normally in-line with the nozzle. By attaching it backwards, I can test with a probe offset of approximately (-30, 15).

1. Build firmware with probe offset of (0,0), and install probe in-line with nozzle
1. M502, M500, Reboot
1. Perform G33 in original configuration
    - Probes in a centered circle, about 10mm from edge of bed
1. Perform G29 in original configuration
    - Even distribution of probe points
1. Reverse probe mounting, and configure using `M851 X-30 Y15`
1. Perform G33
    - Probes in centered circle, with reduced size to account for probe offset
    - (Prior to these changes this crashed badly, potentially damaging the probe)
1. Perform G29
    - Probe distributed across bed, but now skewed to one side, since the bed is not evenly reachable.
1. M500, Reboot
1. Verify that G33 and G29 still work, using the previously saved probe offset

### Effect of changes on example configurations

I modified all delta examples to auto-calculate the DELTA_CALIBRATION_RADIUS.
The following table shows the changes from their old hard-coded values to the new auto-calculated values. Note that for the two FLSUN examples, I reduced MIN_PROBE_EDGE to allow for a DELTA_CALIBRATION_RADIUS resembling the old value. They already had DELTA_AUTO_CALIBRATE enabled, so this seemed to be a safe change.

![image](https://user-images.githubusercontent.com/20053467/68743840-257fd280-05a8-11ea-9b13-828d302ace13.png)

### Related Issues

- #15824 (This was already closed, but without actually fixing it)
- #15790 (Doesn't sound related, but is fixed in this branch along with other probing changes)

I suspect more issues have not been reported for the following reasons:
- Many probes are in-line with the nozzle on delta printers
- Some of these issues were only recently introduced along with @InsanityAutomation's improvement to allow configuring probe offsets through M851.
